Enrich resources with additional metadata

rosstimothy opened the pull request:
Updates ListResources and ListUnifiedResources to optionally allow responses to include the allowed logins per returned resource that requesting user has access to given there roles. Logins are only currently populated for SSH and WindowsDesktop resources. The new `types.EnrichedResource` was added to facilitate transporting the underlying resource and the additional user specific information to consumers.

avatus commented on api/client/client.go:
```diff
@@ -3608,13 +3608,78 @@ func ListUnifiedResourcePage(ctx context.Context, clt ListUnifiedResourcesClient
+				resource = respResource.GetKubernetesServer()
+			case types.KindUserGroup:
+				resource = respResource.GetUserGroup()
+			case types.KindAppOrSAMLIdPServiceProvider:
+				resource = respResource.GetAppServerOrSAMLIdPServiceProvider()
```
`KindAppOrSAMLIdPServiceProvider` is being removed in favor of just using the `SAMLIdPServiceProvider` type here https://github.com/gravitational/teleport/pull/38709

avatus commented on api/client/client.go:
```diff
@@ -3608,13 +3608,78 @@ func ListUnifiedResourcePage(ctx context.Context, clt ListUnifiedResourcesClient
+
+	// Set the limit to the default size if one was not provided within
+	// an acceptable range.
+	if req.Limit <= 0 || req.Limit > int32(defaults.DefaultChunkSize) {
+		req.Limit = int32(defaults.DefaultChunkSize)
+	}
```
does not require a change because this is how we are currently doing it:

Setting the limit to default if not supplied is great. 
Overwriting if it's greater tho might be confusing if I'm building a caller that receives "successful" responses with too large of a limit. It might be better to return an error and say "hey thats too much" so I know that my request is shaped wrong.

Again, don't need to change here. I'll inspect how other APIs handle this case

rosstimothy committed:
Enrich resources with additional metadata

Updates ListResources and ListUnifiedResources to optionally allow
responses to include the allowed logins per returned resource that
requesting user has access to given there roles. Logins are only
currently populated for SSH and WindowsDesktop resources. The new
types.EnrichedResource was added to facilitate transporting the
underlying resource and the additional user specific information to consumers.

rosstimothy committed:
Merge branch 'master' into tross/list_with_logins
